aten/src/ATen/native/cpu/SoftMaxKernel.cpp

```diff
@@ -206,6 +207,268 @@ struct vec_host_softmax_lastdim {
+              Vec output_m256_o2 = output_list[work_idx*2+1]/sum_m256_o2;
+
+              Vec_bf16 output_m256_bf16 = convert_float_bfloat16(output_m256_o1, output_m256_o2);
+              output_m256_bf16.store(output_data);
+            }
+            idx += 16;
+          } else {
+            // Tail cases
+            // There are 2 tail cases:
+            // Case 1. The final tail case of this thread when not enough numbers for parallization
+            // Case 2. For tail case of each inner_size when not enough numbers for parallization
+            int64_t tail_number = 0;
+            if ((idx+16) > end) {
+              // Case1
+              tail_number = end - idx;
+            } else { // (inner_idx+16) > inner_size
+              // Case2
+              tail_number = inner_size - inner_idx;
+            }
+
+            for (int64_t i=0; i < tail_number; i++) {
+
+              int64_t outer_idx = (idx + i) / inner_size;
+              int64_t inner_idx = (idx + i) % inner_size;
+
+              BFloat16* base_input_data =
+                  input_data_base + outer_idx * outer_stride + inner_idx;
+              BFloat16* base_output_data =
+                  output_data_base + outer_idx * outer_stride + inner_idx;
+
+              // Step1: Get max score
+              BFloat16 max_input = base_input_data[0];
+              for (int64_t work_idx = 1; work_idx < dim_size; work_idx += 1) {
+                BFloat16* input_data = base_input_data + work_idx * dim_stride;
+                max_input = std::max(max_input, input_data[0]);
+              }
+
+              // Step2: Calculate the Sum
+              BFloat16 sum_data = 0;
+              for (int64_t work_idx = 0; work_idx < dim_size; work_idx += 1) {
+                BFloat16* input_data = base_input_data + work_idx * dim_stride;
+                BFloat16* output_data = base_output_data + work_idx * dim_stride;
+                output_data[0] = std::exp(input_data[0] - max_input);
+                sum_data += output_data[0];
+              }
+
+              // Step3: Unify
+              for (int64_t work_idx = 0; work_idx < dim_size; work_idx += 1) {
+                BFloat16* output_data = base_output_data + work_idx * dim_stride;
+                output_data[0] = output_data[0]/sum_data;
+              }
+            }
+            idx += tail_number;
+          }
+        }
+      });
+}
+#endif
+
+template <typename scalar_t>
+inline void _vec_softmax(
+    scalar_t* input_data_base,
+    scalar_t* output_data_base,
+    int64_t outer_size, // for example: 32
+    int64_t inner_size, // for example: 15130
+    int64_t dim_size) { // for example: 81
+  using Vec = vec::Vectorized<scalar_t>;
+  int64_t grain_size = internal::GRAIN_SIZE / (16 * dim_size);
+  if (grain_size < 1)
+    grain_size = 1;
+
+  int64_t dim_stride = inner_size; // for example: 15130
+  int64_t outer_stride = dim_size * dim_stride; // for example: 81 * 15130
+
+  parallel_for(
```

Comment:
outer_size and inner_size are not adjacent dimensions, we should not collapse non-adjacent dimensions, otherwise it will result into non-contiguous memory access.

Reply:
How about we only parallel_for along the `inner_size`? And there will be a outer for loop(serial) along the `outer_size`?